### PR TITLE
Trace middleware: replace `exit.case` attribute with `error.type`

### PR DESCRIPTION
**Pull request by iRevive**

### CODE REVIEW

File: core/src/test/scala/org/http4s/otel4s/middleware/ClientMiddlewareTests.scala [hunk 1]
```diff
@@ -95,7 +95,6 @@ class ClientMiddlewareTests extends CatsEffectSuite {
           assertEquals(getAttr[Long]("http.response.status_code"), Some(200L))
           assertEquals(getAttr[Seq[String]]("http.response.header.foo"), None)
           assertEquals(getAttr[Seq[String]]("http.response.header.baz"), Some(Seq("qux")))
-          assertEquals(getAttr[String]("exit.case"), Some("succeeded"))
```
Comment (NthPortal) [inline core/src/test/scala/org/http4s/otel4s/middleware/ClientMiddlewareTests.scala:98]: should we add an assert for status somewhere?

**Review by NthPortal — APPROVED** [review]

File: core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala [hunk 2]
```diff
@@ -151,18 +158,33 @@ object ClientMiddleware {
...
-              span.addAttribute(CustomAttributes.exitCase(ec.toOutcome))
+            resp <- poll(client.run(newReq).mapK(trace)).guaranteeCase {
+              case Outcome.Succeeded(fa) =>
+                fa.evalMap { resp =>
+                  val out = response(resp, allowedResponseHeaders) ++
+                    additionalResponseAttributes(resp.responsePrelude)
+
+                  span.addAttributes(out) >> span
+                    .setStatus(StatusCode.Error)
+                    .unlessA(resp.status.isSuccess)
```
Comment (iRevive) [inline core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala:169]: We should set the status to `Error` if error code is 4xx-5xx - https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-client:

> If the request fails with an error before response status code was sent or received, error.type SHOULD be set to exception type (its fully-qualified class name, if applicable) or a component-specific low cardinality error identifier.
> 
> If response status code was sent or received and status indicates an error according to [HTTP span status definition](https://opentelemetry.io/docs/specs/semconv/http/http-spans/), error.type SHOULD be set to the status code number (represented as a string), an exception type (if thrown) or a component-specific error identifier.
> 
> The error.type value SHOULD be predictable and SHOULD have low cardinality. Instrumentations SHOULD document the list of errors they report.
> 
> The cardinality of error.type within one instrumentation library SHOULD be low, but telemetry consumers that aggregate data from multiple instrumentation libraries and applications should be prepared for error.type to have high cardinality at query time, when no additional filters are applied.
> 
> If the request has completed successfully, instrumentations SHOULD NOT set error.type.

File: core/src/main/scala/org/http4s/otel4s/middleware/ServerMiddleware.scala [hunk 3]
```diff
@@ -188,27 +193,23 @@ object ServerMiddleware {
...
+                    .guaranteeCase {
+                      case Outcome.Succeeded(fa) =>
+                        fa.flatMap { resp =>
+                          val out =
+                            response(
+                              resp,
+                              allowedResponseHeaders,
+                            ) ++ additionalResponseAttributes(resp.responsePrelude)
+
+                          span.addAttributes(out) >> span
```
Comment (iRevive) [inline core/src/main/scala/org/http4s/otel4s/middleware/ServerMiddleware.scala:205]: The same story here

**Comment by iRevive** [comment]
Hey @NthPortal. 

While the `exit.case` doesn't exist in the spec, we should fill `error.type` based on the following cases:
1) If the response was sent and the response code was within 4xx-5xx - the status code must be used
2) If the response cannot be generated due to an error - the error name must be used

File: core/src/test/scala/org/http4s/otel4s/middleware/ClientMiddlewareTests.scala [hunk 4]
```diff
@@ -77,9 +78,10 @@ class ClientMiddlewareTests extends CatsEffectSuite {
           val span = spans.head
           assertEquals(span.name, "Http Client - GET")
           assertEquals(span.kind, SpanKind.Client)
+          assertEquals(span.status, StatusData.Unset)
```
Comment (iRevive) [inline core/src/test/scala/org/http4s/otel4s/middleware/ClientMiddlewareTests.scala:81]: https://opentelemetry.io/docs/specs/semconv/http/http-spans/#status

> [Span Status](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.35.0/specification/trace/api.md#set-status) MUST be left unset if HTTP status code was in the 1xx, 2xx or 3xx ranges, unless there was another error (e.g., network error receiving the response body; or 3xx codes with max redirects exceeded), in which case status MUST be set to Error.

File: core/src/main/scala/org/http4s/otel4s/middleware/TypedAttributes.scala [hunk 5]
```diff
@@ -86,6 +87,10 @@ object TypedAttributes {
   def userAgentOriginal(userAgent: `User-Agent`): Attribute[String] =
     UserAgentAttributes.UserAgentOriginal(`User-Agent`.headerInstance.value(userAgent))
 
+  /** @return the `error.type` `Attribute` */
+  def errorType(tpe: String): Attribute[String] =
```
Comment (iRevive) [inline core/src/main/scala/org/http4s/otel4s/middleware/TypedAttributes.scala:91]: The `String` fills a bit loose. Should I provide two alternatives? 
```scala
def errorType(status: Status): Attribute[String] = ...
def errorType(error: Throwable): Attribute[String] = ...
```

File: core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala [hunk 6]
```diff
@@ -227,6 +241,13 @@ object ClientMiddleware {
     builder ++= TypedAttributes.Headers.response(response.headers, allowedHeaders)
     builder ++= response.attributes.lookup(ExtraAttributesKey).toList.flatten
 
+    // https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-client
+    // [5]: If response status code was sent or received and status indicates an error according
+    // to HTTP span status definition, `error.type` SHOULD be set to the status code number (represented as a string),
+    // an exception type (if thrown) or a component-specific error identifier.
+    if (!response.status.isSuccess)
+      builder += TypedAttributes.errorType(response.status.code.toString)
```
Comment (NthPortal) [inline core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala:249]: is there a reason not to do this up above where you set the span status?

File: core/src/main/scala/org/http4s/otel4s/middleware/TypedAttributes.scala (hunk 5 above)
Comment (NthPortal) [reply core/src/main/scala/org/http4s/otel4s/middleware/TypedAttributes.scala:91]: I think so, yeah

File: core/src/test/scala/org/http4s/otel4s/middleware/ClientMiddlewareTests.scala [hunk 7]
```diff
@@ -177,20 +178,28 @@ class ClientMiddlewareTests extends CatsEffectSuite {
             }
 
             val tracedClient = ClientMiddleware.default[IO].build(fakeClient)
-            val request = Request[IO](Method.GET, uri"http://localhost/?#")
```
Comment (NthPortal) [inline core/src/test/scala/org/http4s/otel4s/middleware/ClientMiddlewareTests.scala:180]: what's the reason for this change?

File: core/src/main/scala/org/http4s/otel4s/middleware/TypedAttributes.scala (hunk 5 above)
Comment (NthPortal) [reply core/src/main/scala/org/http4s/otel4s/middleware/TypedAttributes.scala:91]: I think so, yeah

File: core/src/test/scala/org/http4s/otel4s/middleware/ClientMiddlewareTests.scala (hunk 7 above)
Comment (iRevive) [reply core/src/test/scala/org/http4s/otel4s/middleware/ClientMiddlewareTests.scala:180]: Fewer attributes to test. `url.query` and `url.fragment` will be set to an empty string anyway.
And we test that `url.query` and `url.fragment` are extracted correctly above.

File: core/src/main/scala/org/http4s/otel4s/middleware/TypedAttributes.scala (hunk 5 above)
Comment (iRevive) [reply core/src/main/scala/org/http4s/otel4s/middleware/TypedAttributes.scala]: Done

File: core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala (hunk 6 above)
Comment (iRevive) [reply core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala]: Two motivational reasons:
1) Keep `ClientMiddleware.response` and `ServerMiddleware.response` looking similar
2) Use `span.addAttributes` only once

I don't mind moving it to the [guaranteeCase](https://github.com/http4s/http4s-otel4s-middleware/blob/96b0da548bba78ca171666b3ff057917533d01ad/core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala#L167-L169).

**Review by NthPortal — APPROVED** [review]
@iRevive ready for merge?

File: core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala (hunk 6 above)
Comment (NthPortal) [reply core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala]: > I don't mind moving it to the [guaranteeCase](https://github.com/http4s/http4s-otel4s-middleware/blob/96b0da548bba78ca171666b3ff057917533d01ad/core/src/main/scala/org/http4s/otel4s/middleware/ClientMiddleware.scala#L167-L169).

nah, don't worry about it. thank you for the explanation 👍

**Comment by iRevive** [comment]
> @iRevive ready for merge?

Yep, it can be merged. Thanks for the review.
